buildpackage/builder.go

```diff
@@ -0,0 +1,138 @@
+package buildpackage
+
+import (
+	"io/ioutil"
+	"os"
+
+	"github.com/buildpack/imgutil"
+	"github.com/pkg/errors"
+
+	"github.com/buildpack/pack/dist"
+	"github.com/buildpack/pack/style"
+)
+
+type ImageFactory interface {
+	NewImage(repoName string, local bool) (imgutil.Image, error)
+}
+
+type PackageBuilder struct {
+	defaultBpInfo dist.BuildpackInfo
+	buildpacks    []dist.Buildpack
+	stacks        []dist.Stack
+	imageFactory  ImageFactory
+}
+
+func NewBuilder(imageFactory ImageFactory) *PackageBuilder {
+	return &PackageBuilder{
+		imageFactory: imageFactory,
+	}
+}
+
+func (p *PackageBuilder) SetDefaultBuildpack(bpInfo dist.BuildpackInfo) {
+	p.defaultBpInfo = bpInfo
+}
+
+func (p *PackageBuilder) AddBuildpack(buildpack dist.Buildpack) {
+	p.buildpacks = append(p.buildpacks, buildpack)
+}
+
+func (p *PackageBuilder) AddStack(stack dist.Stack) {
+	p.stacks = append(p.stacks, stack)
+}
+
+func (p *PackageBuilder) Save(repoName string, publish bool) (imgutil.Image, error) {
+	if err := validateDefault(p.buildpacks, p.defaultBpInfo); err != nil {
+		return nil, err
+	}
+
+	if err := validateStacks(p.buildpacks, p.stacks); err != nil {
+		return nil, err
+	}
+
+	image, err := p.imageFactory.NewImage(repoName, !publish)
+	if err != nil {
+		return nil, errors.Wrapf(err, "creating image")
+	}
+
+	if err := dist.SetLabel(image, MetadataLabel, &Metadata{
+		BuildpackInfo: p.defaultBpInfo,
+		Stacks:        p.stacks,
+	}); err != nil {
+		return nil, err
+	}
+
+	tmpDir, err := ioutil.TempDir("", "create-package")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	for _, bp := range p.buildpacks {
+		bpLayerTar, err := dist.BuildpackLayer(tmpDir, 0, 0, bp)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := image.AddLayer(bpLayerTar); err != nil {
+			return nil, errors.Wrapf(err, "adding layer tar for buildpack %s", style.Symbol(bp.Descriptor().Info.ID+"@"+bp.Descriptor().Info.Version))
+		}
+	}
+
+	if _, err := image.Save(); err != nil {
+		return nil, err
+	}
+
+	return image, nil
+}
+
+func validateDefault(bps []dist.Buildpack, defBp dist.BuildpackInfo) error {
+	if defBp.ID == "" || defBp.Version == "" {
+		return errors.New("a default buildpack must be set")
+	}
+
+	if !bpExists(bps, defBp) {
+		return errors.Errorf("selected default %s is not present",
+			style.Symbol(defBp.ID+"@"+defBp.Version),
+		)
+	}
+
+	return nil
+}
+
+func validateStacks(bps []dist.Buildpack, stacks []dist.Stack) error {
```

Comment:
This is probably a larger question, but I wonder why we even need to declare stacks in `package.toml`. They seem to be computable on the fly as the intersection of all stack sets from each buildpack. cc: @sclevine, @ekcasey